unix,win: add uv_os_uname(), uv_os_free_utsname()

cjihrig opened the pull request:
Notes:

- ~~The Windows implementation isn't written yet in this PR.~~
- ~~In Node's `os` module, the `sysname` field is always `'Windows_NT'` on non-POSIX platforms. Should we mirror that here? Probably not, but it could impact the integration process in Node.~~
- ~~In Node's `os` module on AIX and Windows, the `release` field has a custom implementation.~~
- ~~CI run (https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1133/) shows issues on SmartOS (looks like SmartOS returns a non-negative value on success, where other platforms return 0, will test).~~
- ~~Latest CI run: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1135/ (no related failures except missing Windows implementation)~~

Fixes: https://github.com/libuv/libuv/issues/2126

cjihrig commented:
This is now implemented on all platforms and is ready for review. Some questions that I'd like feedback on, especially on the Windows side:

- I think the implementation could be more like `uname(3)` and drop `uv_os_free_utsname()` if `uv_utsname_t` used arrays instead of `char*`s. It would just require making some guesses on the array sizes. I'm not sure if it's safe to do, but would simplify things quite a bit.
- Are the fields on Windows good in general?
- Should Windows, Cygwin, and MINGW32 have identical output?
- The Windows version uses `GetVersionExW()`, which is deprecated. That doesn't give me a great feeling.

CI with no related failures: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1149/

bnoordhuis commented on src/win/util.c:
```diff
@@ -1627,3 +1628,147 @@ int uv_os_setpriority(uv_pid_t pid, int priority) {
+           (unsigned int) os_info.dwMajorVersion,
+           (unsigned int) os_info.dwMinorVersion);
+#else
+  strcpy(buffer->sysname, "Windows_NT");
```
`uv__strscpy()`, please. :-) Also on  lines 1730 and down.

cjihrig commented:
Thanks, Ben. I've moved this to using arrays and addressed the other feedback. CI of the latest with no related failures: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1152/

cjihrig commented:
@bzoz, @refack do either of you have any input on the Windows design? Also cc @piscisaureus in case you'd like to take a look.

bzoz commented:
The Windows part is basically nonfunctional. `GetVersionEx` will always return Win 8.1. In the CI  it is always (6.2.something as release). 

I had somewhere some code that detects Windows version, I'll look for it.

bzoz commented:
Ok, found it: in https://github.com/libuv/libuv/pull/1346 I've used [RtlGetVersion](pRtlGetVersion), which returns proper Windows version numbers

cjihrig commented:
@bzoz thanks. I tried to use `RtlGetVersion()`, but couldn't get it working. What do you think about trying to use `RtlGetVersion()`, and falling back to `GetVersionEx()` if it isn't available? Conveniently, they both use `OSVERSIONINFOW`.

bzoz commented:
SGTM

cjihrig commented:
Updated to try using `RtlGetVersion()` before falling back to `GetVersionEx()` on Windows. New, good CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1153/.

bnoordhuis commented on src/win/util.c:
```diff
@@ -1627,3 +1627,126 @@ int uv_os_setpriority(uv_pid_t pid, int priority) {
+               (unsigned int) os_info.dwMajorVersion,
+               (unsigned int) os_info.dwMinorVersion,
+               (unsigned int) os_info.dwBuildNumber);
+  if (r >= sizeof(buffer->release)) {
```
This is logically impossible. :-) I'd just turn it into an assert or even a STATIC_ASSERT.

(The maximum length is `2 + 3 * round(log10((unsigned int) -1)) == 32`.)

ghesketh commented:
It feels to me like the intent of the Single UNIX Specification for uname() would mean returning results like the following:

sysname: "Windows Server 2016 Standard"
release: "6.3"
version: "1607"

As you probably know or suspect, these are available as (respectively) ProductName, CurrentVersion, and ReleaseId below HKLM\SOFTWARE\Microsoft\Windows NT\CurrentVersion in the registry.

cjihrig commented:
Looking at https://en.wikipedia.org/wiki/Uname, I don't think the product name is the right thing to place in `sysname`. I'd be OK with combining the product name and service pack information in `version` though. Ultimately, I'll defer to the previously @-mentioned Windows experts.

ghesketh commented:
Hoping not to be tiresome, I thought it might be worth a little expansion:  ProductName made sense to me since "Windows NT" (or Windows_NT in the environment) is the name of the operating system (uname -o), whereas uname -s wants "the name of the implementation of the operating system" for the sysname parameter.

Having said that, I lack knowledge of the intended use cases for uv_os_uname(), so I am happy to defer to parties with specific expectations.

Cheers!

cjihrig commented:
Another thing I should have mentioned is that "Windows_NT" is used by both [Node.js](https://github.com/nodejs/node/blob/b83088b0bb2b8dfa6ee72a807cc10fe1ac898278/src/node_os.cc#L98), as well as [this](https://github.com/gagern/gnulib/blob/6465306025facca3881bcea411b757b0fe8e8df8/lib/uname.c#L110-L122) implementation. I think if we include the product name would make the most sense in the `version` field like [this](https://github.com/gagern/gnulib/blob/6465306025facca3881bcea411b757b0fe8e8df8/lib/uname.c#L150-L157) (platforms like [macOS](https://ci.nodejs.org/job/libuv-test-commit-osx/1167/nodes=osx1011/console) seem to use `version` for longer text as well).

kadler commented:
On AIX and IBM i, the machine information returned by uname is not very useful: it returns the machine serial number and some other machine attributes. eg.

```
OS400 wernstrom 2 7 0010000ACBAA Os
AIX hoss 1 7 00F601E54C00 powerpc AIX
```

For IBM i, we could hard-code this to ppc64 or similar. For AIX, there is probably some API that can be called to determine the architecture, but realistically it could be hardcoded to ppc64 as well, since AFAIK that's the only platform it currently runs on.

@libuv/ibmi @libuv/aix

richardlau commented:
> 
> 
> On AIX and IBM i, the machine information returned by uname is not very useful: it returns the machine serial number and some other machine attributes. eg.
> 
> ```
> OS400 wernstrom 2 7 0010000ACBAA Os
> AIX hoss 1 7 00F601E54C00 powerpc AIX
> ```
> 
> For IBM i, we could hard-code this to ppc64 or similar. For AIX, there is probably some API that can be called to determine the architecture, but realistically it could be hardcoded to ppc64 as well, since AFAIK that's the only platform it currently runs on.
> 
> @libuv/ibmi @libuv/aix

If we hard-code for IBM i I'd be okay with doing the same for AIX if it makes the code changes more compact.

cjihrig commented:
@richardlau and @kadler: just to clarify, you are fine with the `machine` field on AIX and IBM i being hard coded to `'ppc64'`? I'm looking to land this soon and wanted to double check.

ThePrez commented:
I also agree with `ppc64`

cjihrig commented:
CI with updates for AIX and IBM i: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1168/. Still looks like no related failures.
